### PR TITLE
Handle smooth_frames=False consistently for single and multi-frequency records

### PR DESCRIPTION
**Pull request by bemoody**
The `smooth_frames` argument to `rdrecord` is meant to determine whether the input signals are stored as `{d|p}_signal` (a homogeneous two-dimensional array, requiring all signals to be resampled to the same frequency) or `e_{d|p}_signal` (a list of one-dimensional arrays, keeping each signal at its actual sampling frequency).

However, this *only* works at present if you are trying to read signals with multiple samples per frame.  (The assumption, I guess, being that if the record only has one sample per frame, then smooth format doesn't discard any information, so there is no reason for the application to want non-smooth format.)

IMNSHO, this is not a good idea: the application should not need to know in advance whether the record contains multiple samples per frame, and should not need to implement two different logic branches, simply to be able to read input signals without resampling.  If the application asks for a 2D array it should get a 2D array, and if the application asks for a list of 1D arrays it should get a list of 1D arrays.

Fixing this could break applications, because if an existing application only ever reads single-frequency records, and it specifies `smooth_frames=False` and `physical=True`, it would receive `p_signal` data just as if it had specified `smooth_frames=True`.  It would be silly for an application to do that, and sillier to rely on this behavior, but who knows.

This fixes issue #312 and also fixes the handling of `smooth_frames=False` together with `sampfrom`/`sampto`.

### CODE REVIEW

**Comment by briangow** [comment]
@bemoody this appears to be working from my testing. I do think we could improve some of the descriptions about how this works though. It took a fair bit of digging to find a good definition for how `smooth_frames` works and to find a database which uses a variable number of `samps_per_frame` ([drivedb](https://physionet.org/content/drivedb/1.0.0/) is one such database). 

In addition to describing that `smooth_frames = True` will return a uniformly sampled array and `smooth_frames = False` will return a list of signals at their original sample rates, we could be clearer about how the "smoothing" works. Perhaps by stating that `smooth_frames = True` will decimate the signal by its `samps_per_frame` in the parameter descriptions. 

In anticipation of merging this update I've created a new issue to reflect that we need to update the wfdb-python help documentation since it currently states that `smooth_frames` is only used when signals have multiple samples per frame, see #316 .

**Comment by bemoody** [comment]
Another issue is that the current implementation of `smooth_frames=False` is very slow, much slower than `smooth_frames=True`.

(For example, on my machine `wfdb.rdrecord("sample-data/100")` is nearly instantaneous, while `wfdb.rdrecord("sample-data/100", smooth_frames=False) takes around 8 seconds.)

I'm not sure why this is.

**Comment by bemoody** [comment]
@briangow, if I understand correctly, you're saying we should change
the name of the argument for _rd_segment and for _rd_dat_signals, since
in those two cases the type of the return value is dependent on the
argument.  I agree that would be clearer and can certainly make that
change.

(I disagree, though, that it isn't feasible for these functions to
always return the same type.  In fact, we could change both functions
to *always* return a list of arrays, and post-process the result in
rdrecord.  However, as this would be a significant change code-wise,
and one with no user-visible effects, I would propose to do this in a
later pull request.)

The question I'm still unsure about, which wasn't completely clear from
Chen's message, is whether we also need to make a change in the *public
API* (rdrecord).  For rdrecord, 'smooth_frames' doesn't change the type
of the result (it's always a Record object); it only affects whether the
'd_signal' or 'e_d_signal' attribute is set.

**Comment by briangow** [comment]
While I think it would be nice to have a clearer name for this parameter in the public API, it might cause more of a mess than it's worth. I think it would be more effective to focus on improving the documentation and tutorials/demo packages.

File: wfdb/io/record.py
Comment (tompollard) [reply wfdb/io/record.py]: Looks like this is resolved.

**Comment by tompollard** [comment]
@cx1111 please could you respond to this point?:

> The question I'm still unsure about, which wasn't completely clear from
Chen's message, is whether we also need to make a change in the *public
API* (rdrecord).  For rdrecord, 'smooth_frames' doesn't change the type
of the result (it's always a Record object); it only affects whether the
'd_signal' or 'e_d_signal' attribute is set.

**Comment by cx1111** [comment]
I agree that the current API which requires the caller to know the frequency of the signals, is poorly suited for an application writer trying to process large volumes of data. I think when I designed this, that I only thought about users trying to analyze a small homogeneous datasets they are familiar with.

> rdrecord could *always* set e_d_signal/e_p_signal, and d_signal/p_signal could be computed lazily on request; i.e., the smooth_frames argument could be deprecated and ignored completely 

Lazy evaluation sounds interesting and scary :P This would probably require a lot of further thought on the design and tradeoffs. Perhaps for the future.

> Just to be clear, though, are you saying this should be changed in the public API (rdrecord)? Or only the internal API of wfdb.io._signal?

Internally, we should **definitely** add more params to control the behavior of `_rd_dat_signals`. In particular, this [if else](https://github.com/MIT-LCP/wfdb-python/blob/0d416032641745a83feea406b300035c424e20d4/wfdb/io/_signal.py#L1215) statement is rather nasty.

Perhaps we could add another function parameter such as `return_dims` which specifies the dimensions of the output array(s), which explicitly causes the function to return either the list of 1d arrays (val=1), or the 2d array (val=2). Then we can have the following behavior matrix with `return_dims` and `smooth_frames`:
- (1, True): List of 1d arrays. Guaranteed to be same length.
- (1, False): List of 1d arrays. Not guaranteed to be same length.
- (2, True): Single 2d array.
- (2, False): Illegal combination.

We could also add this new param to `rdrecord` and use the same behavior matrix.

This way, the user can specify both the desired output data type and the behavior for multi-frequency records. the `smooth_frames` param could be safely ignored for single-frequency records.

The changes could then be made in 2 steps:
1. Change the internal signal functions.
2. Change the public facing API of rdrecord.

Thoughts?

**Comment by bemoody** [comment]
Thanks!

To address your "step 1", I think the most straightforward solution is to remove the "smoothing" functionality from `_rd_segment` and `_rd_dat_signals` entirely.  Perhaps that sounds like a huge, dramatic change, but it isn't really. :)  The function `Record.smooth_frames` already exists, and it should give numerically identical results to what `_rd_dat_signals` does at present.

Aside from reducing the amount of duplicated logic, the implementation of `Record.smooth_frames` is also better designed (IMHO) and probably more efficient overall.  (There are some slight drawbacks to "unpacking" and later "repacking" the array, but I think this is a minor issue and outweighed by the benefit of simplifying the codebase.)  The performance can be further improved in later pull requests.

Note that 9d77075ebff003222c7edab84c94e6e6289481fc is 16fa5ed70209c63fadda5541e8fd6395ef02ef43 rebased onto 0d416032641745a83feea406b300035c424e20d4.  The new commits here are ca20b1a360b10315f2ec5eab6d2df382497b8583 ("rdrecord: smooth frames by invoking smooth_frames()"), b2f39955f0ba42e9f422ddf409dbc0bb772bf226 ("_rd_segment, _rd_dat_signals: remove support for smooth_frames=True"), and d02c478be6b2ae753eaddb53175cc7d147a328ae ("_rd_segment, _rd_dat_signals: remove smooth_frames parameter").

**Comment by bemoody** [comment]
One more thing I'll say is that this restructuring should *also* make it much easier to add support for other signal file formats (notably WFDB-FLAC, but also to properly handle EDF/BDF, and maybe other things in the future.)

**Comment by cx1111** [comment]
Will try to review by EOW. Ping me if I forget.

**Review by cx1111 — APPROVED** [review]
Looks good. Nice code deletion. Perhaps change the title of the PR and edit the OP description to include a summary of the changes?

**Comment by cx1111** [comment]
Also from now on, repository owners should merge their own PRs once they have approval.

**Comment by bemoody** [comment]
Hopefully the last set of changes in this pull request. :)

Changing `smooth_frames()` to return the minimal data type requires changing the logic a little (we can't *just* change the type of the array because the intermediate sums can easily exceed the range of the original data type.)  At the same time, throw in a few optimizations.

Speed comparison:

(master - original logic of `_rd_dat_signals`)
```
>>> import timeit
>>> timeit.timeit('wfdb.rdrecord("sample-data/03700181")', 'import wfdb', number=100)
54.588751977000356
```

(d02c478b - original logic of `smooth_frames`)
```
>>> import timeit
>>> timeit.timeit('wfdb.rdrecord("sample-data/03700181")', 'import wfdb', number=100)
1.1534926320018712
```

(non-smooth - `smooth_frames` with further optimizations)
```
>>> import timeit
>>> timeit.timeit('wfdb.rdrecord("sample-data/03700181")', 'import wfdb', number=100)
1.0475175969986594
```

**Review by cx1111 — APPROVED** [review]
Good catch, and nice job with the improvements!

What exactly is responsible for most of the speedup? The smaller memory allocation chunks, or something else?

**Pull request merged**
